[MRG+1] Change VotingClassifier estimators by set_params

yl565 opened the pull request:
PR to #7288. Continuation of #7484

yl565 committed:
PR to 7288
Use _BaseComposition as base

yl565 committed:
Fix flakes problem

jnothman commented:
You shouldn't need to open a new PR for a rebase. I'll try take a look at this over the coming weeks. Sorry for the slow reviews

yl565 commented:
@jnothman, I have just started using git, do you mind letting me know if the following procedure is correct for a rebase (assume origin is already up to date with upstream/master):

```
git checkout mybranch
git rebase master
git push origin master
```

amueller commented:
the last should be ``git push origin mybranch``.
Also see : http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#rebasing-on-master

jnothman commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -44,7 +43,8 @@ class VotingClassifier(BaseEstimator, ClassifierMixin, TransformerMixin):
         Invoking the ``fit`` method on the ``VotingClassifier`` will fit clones
         of those original estimators that will be stored in the class attribute
-        `self.estimators_`.
+        `self.estimators_`. An estimator can be set to `None` using
```
these should use double-backticks.

jnothman commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -152,6 +152,11 @@ def fit(self, X, y, sample_weight=None):
+        isnone = np.array([1 if clf is None else 0
+                           for _, clf in self.estimators])
+        if isnone.sum() == len(self.estimators):
+            raise ValueError('All estimators is None. At least one is required'
```
'is None' -> 'are None'

jnothman commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -161,11 +166,19 @@ def fit(self, X, y, sample_weight=None):
 
         self.estimators_ = Parallel(n_jobs=self.n_jobs)(
```
This change to `estimators_` needs to be documented under `Attributes`

jnothman commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -161,11 +166,19 @@ def fit(self, X, y, sample_weight=None):
         return self
 
+    @property
+    def _narej_weights(self):
```
What is narej?

jnothman replied:
Confusing because there aren't any NaNs involved. How about `_weights_not_none`... or just remove this helper

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -7,10 +7,54 @@
 __all__ = ['if_delegate_has_method']
 
 
+class _BaseComposition(six.with_metaclass(ABCMeta, BaseEstimator)):
```
You should be using this in `Pipeline` and `FeatureUnion`

yl565 replied:
Do you mean I should also modify Pipeline and FeatureUnion to use _BaseComposition?

jnothman replied:
yes

jnothman commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -258,3 +258,49 @@ def test_sample_weight():
+    eclf2.set_params(voting='soft').fit(X, y)
+    assert_array_equal(eclf1.predict(X), eclf2.predict(X))
+    assert_array_equal(eclf1.predict_proba(X), eclf2.predict_proba(X))
+    msg = ('All estimators is None. At least one is required'
```
is -> are

jnothman commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -258,3 +258,49 @@ def test_sample_weight():
+    eclf1.set_params(voting='soft').fit(X, y)
+    eclf2.set_params(voting='soft').fit(X, y)
+    assert_array_equal(eclf1.predict(X), eclf2.predict(X))
+    assert_array_equal(eclf1.predict_proba(X), eclf2.predict_proba(X))
```
Please test soft `transform`. The outputs should differ between the 0-weight and None variants, though...

jnothman reviewed (changes requested):
Sorry this has taken a while to get to

yl565 committed:
Change ``pipeline``, add more tests and other changes
1. Use ``_BaseComposition`` in class ``Pipeline`` and ``FeatureUnion``
2. Add tests of soft voting ``transform`` when one estimator is set to None
3. Add estimator name validation in ``_BaseComposition`` and tests
4. Other requested changes.

yl565 commented:
@jnothman I've made the requested changes. Also add estimator name validation and tests.

jnothman commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -258,3 +271,63 @@ def test_sample_weight():
+    eclf2 = VotingClassifier(estimators=[('rf', clf2), ('nb', clf3)],
+                             voting='soft', weights=[1, 0.5])
+    eclf2.set_params(rf=None).fit(X1, y1)
+    assert_array_equal(eclf1.transform(X1), np.array([[[0.7, 0.3], [0.3, 0.7]],
```
Hmm. Looking at this makes me wonder whether we should be multiplying the outputs by the weight. Not an issue for this PR.

jnothman commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -102,11 +103,14 @@ class VotingClassifier(BaseEstimator, ClassifierMixin, TransformerMixin):
         self.n_jobs = n_jobs
 
+    @property
+    def named_estimators(self):
```
I wish this didn't exist, but I know it's not your problem.

jnothman replied:
I wish that we had not copied this bad design feature from pipeline!

jnothman reviewed (approved):
LGTM, thanks!

jnothman commented:
Looking for a reviewer...

jnothman commented:
This also needs merging with updated master.

yl565 committed:
Merge branch 'master' into set_params_VC_7288

yl565 committed:
Remove the unused import warn

yl565 commented:
@jnothman I merged it with updated master. Can you help me check why the appveyor build is cancelled?

jnothman commented:
Build was cancelled because all our appveyors were stuck and were cancelled. Don't worry about it.

prcastro commented:
Anything preventing this from getting merged?

jnothman commented:
> Anything preventing this from getting merged?

A backlog and low reviewer availability. We require two approvals.

jnothman commented:
Feel like reviewing this, @lesteve? @raghavrv?

MechCoder commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -260,6 +273,66 @@ def test_sample_weight():
+    eclf2.set_params(nb=clf2).fit(X, y)
+    assert_array_equal(eclf1.predict(X), eclf2.predict(X))
+    assert_array_equal(eclf1.predict_proba(X), eclf2.predict_proba(X))
+
```
(It might be possible that two different classifiers give the same predictions)

MechCoder reviewed (commented):
Just some minor comments related to testing.

MechCoder commented:
Extremely sorry for the long wait @yl565 ! There seems to be two additions in this PR:

1. Ability to substitute or set a classifier directly using the `set_params` interface.
2. Ability to set the classifier to None or disabling it using `set_params`.

Given that in master, both these fail silently, should we document these as new features or bug-fixes?

jnothman commented:
Likely they don't just fail silently, but set an attribute on the
estimator, even overwriting existing attributes! We had the same for
Pipeline until 0.18. But I still think, in this case, it's better regarded
as an enhancement.

On 4 April 2017 at 14:18, Manoj Kumar <notifications@github.com> wrote:

> Extremely sorry for the long wait @yl565 <https://github.com/yl565> !
> There seems to be two additions in this PR:
>
>    1. Ability to substitute or set a classifier directly using the
>    set_params interface.
>    2. Ability to set the classifier to None or disabling it using
>    set_params.
>
> Given that in master, both these fail silently, should we document these
> as new features or bug-fixes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7674#issuecomment-291389148>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz618f7dT7XLeBL9qTnutuvLH1LBt2ks5rscSLgaJpZM4KXgCa>
> .
>

MechCoder commented:
Ah I see why, in that case we should also test that

```python
vclf.set_params(nb=clf2)
assert_false(hasattr(vclf, "nb")
```

Also, a weird corner case, but it might be a good idea to also test what happens when one sets estimators, the classifiers by themselves and the hyperparameters of the classifiers at once.

MechCoder commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -252,17 +270,13 @@ def transform(self, X):
         else:
             return self._predict(X)
 
+    def set_params(self, **params):
```
Can you document this?

MechCoder commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -252,17 +270,13 @@ def transform(self, X):
-                for key, value in six.iteritems(step.get_params(deep=True)):
-                    out['%s__%s' % (name, key)] = value
-            return out
+        return super(VotingClassifier,
```
Can you document this?

yl565 committed:
Add more test and documentation

yl565 committed:
resolve conflict with master

yl565 commented:
@MechCoder I added the tests and documentation. Could you please explain more on the following two of your comments?

> Can you also check eclf2.estimators_, eclf2.estimators and eclf2.get_params()

> Also,a weird corner case, but it might be a good idea to also test what happens when one sets estimators, the classifiers by themselves and the hyperparameters of the classifiers at once.

MechCoder commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -260,6 +273,66 @@ def test_sample_weight():
+                                         ('nb', clf3)],
+                             voting='hard', weights=[1, 1, 0.5])
+    eclf2.set_params(rf=None).fit(X, y)
+    assert_array_equal(eclf1.predict(X), eclf2.predict(X))
```
I was suggesting to test the behaviour of `eclf2.estimators`, `eclf2.estimators_`  and ``eclf2.get_params()`.

```python
assert_true(dict(eclf2.estimators)["rf"] is None)
assert_true(len(eclf2.estimators_) == 2)
assert_true(all([not isinstance(est, RandomForestClassifier)  for est in eclf2.estimators_])
assert_true(eclf2.get_params()["rf"] is None)
```

MechCoder commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -252,17 +269,29 @@ def transform(self, X):
+    def set_params(self, **params):
+        """ Setting the parameters for the voting classifier
+
+        Valid parameter keys can be listed with get_params().
```
Can you add under the `get_params` heading? I would just say "Get the parameters of the VotingClassifier". In addition, I would also document the parameter deep saying that setting it to True gets the various classifiers and the parameters of the classifiers as well.

MechCoder commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -252,17 +269,29 @@ def transform(self, X):
+            Specific parameters using e.g. set_params(parameter_name=new_value)
+            Estimators can be removed by setting them to None. In the following
+            example, the RandomForestClassifier is removed:
+            clf1 = LogisticRegression()
```
I am fairly sure that you have to add this under a side-heading "Examples" so that it renders properly

MechCoder commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -252,17 +269,29 @@ def transform(self, X):
+        Parameters
+        ----------
+        params: keyword arguments
+            Specific parameters using e.g. set_params(parameter_name=new_value)
```
In addition, to setting the parameters of the ``VotingClassifier``, (doubleticks) the individual classifiers of the ``VotingClassifier`` can also be set or replaced by setting them to None.

MechCoder commented:
That is it from me! Please add a whatsnew under Enhancements and rebase properly so that I can merge.

MechCoder commented:
You just need to do an interactive rebase

```
git rebase -i master
```

It will spit out errors wherever there are merge conflicts. You would need to check wherever the following block is present.

```
<<<<<<< HEAD
some code
========
some code
>>>>>>>>
```

and decide how you would like to merge the two blocks. Then you do 

```
git add name_of_file
git rebase --continue
```

and keep continuing.

jnothman commented:
No, just do git merge upstream/master if you can avoid a rebase. Much
easier.

On 8 April 2017 at 05:52, Manoj Kumar <notifications@github.com> wrote:

> You just need to do an interactive rebase
>
> git rebase -i master
>
> It will spit out errors wherever there are merge conflicts. You would need
> to check wherever the following block is present.
>
> <<<<<<< HEAD
> some code
> ========
> some code
> >>>>>>>>
>
> and decide how you would like to merge the two blocks. Then you do
>
> git add name_of_file
> git rebase --continue
>
> and keep continuing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7674#issuecomment-292635634>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_Z2GITkewQ9le1D3MgleZzJpC23ks5rtpP-gaJpZM4KXgCa>
> .
>

yl565 committed:
Add testing cases and modify documentation

yl565 committed:
Merge remote-tracking branch 'upstream/master' into set_params_VC_7288

yl565 committed:
Merge branch 'master' into set_params_VC_7288

yl565 committed:
Add to whats_new.rst

MechCoder commented:
Travis is failing because of some cosmetic reasons. Could you fix that?

yl565 committed:
Fix too many blank lines

yl565 commented:
Thanks @MechCoder

MechCoder commented:
Thanks you @yl565 !!

prcastro commented:
🎉
